### DISCUSSION TITLE
No Access to ORT Environment (v2)

### DISCUSSION THREAD

**Discussion by aykut-bozkurt** [pykeio/ort#140] [General]
Hello, with the version **v2**, I see we plan to [quit accessing to `Environment`](https://ort.pyke.io/migrating/v2#environment-no-more). Actually, this makes sense since we allow 1 Environment per process. But it has a limitation for the following case:

We have a http service where we need to update the `Environment` if any setting, e.g. intra-op-parallelism and spin-control, changes. With the new version, this functionality goes away since we cannot recreate an `Environment` with new configuration. When we had access to it, we could drop it and recreate a new one in a controlled way (making sure there is no active run etc...).

One alternative might be to expose a function like `environment_set_config`.  I may also think about it more. What do you think @decahedron1?

**Comment by decahedron1** [comment]
Would https://github.com/pykeio/ort/commit/ea7d0597f7b4c35c041e3cc783d0b146be77256c work? It allows `ort::init` to set the global environment after the first initialization. (Sessions will still need to be recreated to use the new environment, of course.)

**Reply by aykut-bozkurt** [reply]
So much thanks. This works almost as expected. Looks like 1 thing is missing. We need to invoke `Drop` of the `Environment` before calling [CreateEnv()](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a22085f699a2d1adb52f809383f475ed1). Otherwise, ONNX Runtime will return [the same environment instance](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a22085f699a2d1adb52f809383f475ed1:~:text=Note-,Invoking%20this%20function%20will%20return%20the%20same%20instance%20of%20the%20environment%20as%20that%20returned%20by%20a%20previous%20call%20to%20another%20env%20creation%20function%3B%20all%20arguments%20to%20this%20function%20will%20be%20ignored.,-Parameters).

After debugging, I found out that adding something like below just before calling `CreateEnv` works
```rust
pub fn commit(self) -> Result<()> {
  unsafe {
    *G_ENV.cell.get() = None; // this will invoke Drop for previous Environment
  };
  
  let env_ptr = if let Some(global_thread_pool) = self.global_thread_pool_options {
  .....

```

**Reply by decahedron1** [reply]
Fixed in https://github.com/pykeio/ort/commit/127154b591f335e328226ae0fe6257bfc7c09da3

**Reply by aykut-bozkurt** [reply]
It works! Thank you.
